### PR TITLE
Fix type annotation in pandas.compat.numpy.function

### LINKED ISSUE #25950: Fix Type Annotations in pandas/compat/numpy/function.py
Part of #25882

Current errors are as follows:

> mypy pandas/compat/numpy/function.py

    pandas\compat\numpy\function.py:104: error: Need type annotation for 'ARGSORT_DEFAULTS'
    pandas\compat\numpy\function.py:105: error: Cannot determine type of 'ARGSORT_DEFAULTS'
    pandas\compat\numpy\function.py:106: error: Cannot determine type of 'ARGSORT_DEFAULTS'
    pandas\compat\numpy\function.py:107: error: Cannot determine type of 'ARGSORT_DEFAULTS'
    pandas\compat\numpy\function.py:108: error: Cannot determine type of 'ARGSORT_DEFAULTS'
    pandas\compat\numpy\function.py:113: error: Need type annotation for 'ARGSORT_DEFAULTS_KIND'
    pandas\compat\numpy\function.py:114: error: Cannot determine type of 'ARGSORT_DEFAULTS_KIND'
    pandas\compat\numpy\function.py:115: error: Cannot determine type of 'ARGSORT_DEFAULTS_KIND'
    pandas\compat\numpy\function.py:116: error: Cannot determine type of 'ARGSORT_DEFAULTS_KIND'
    pandas\compat\numpy\function.py:158: error: Need type annotation for 'COMPRESS_DEFAULTS'
    pandas\compat\numpy\function.py:159: error: Cannot determine type of 'COMPRESS_DEFAULTS'
    pandas\compat\numpy\function.py:160: error: Cannot determine type of 'COMPRESS_DEFAULTS'
    pandas\compat\numpy\function.py:161: error: Cannot determine type of 'COMPRESS_DEFAULTS'
    pandas\compat\numpy\function.py:164: error: Need type annotation for 'CUM_FUNC_DEFAULTS'
    pandas\compat\numpy\function.py:165: error: Cannot determine type of 'CUM_FUNC_DEFAULTS'
    pandas\compat\numpy\function.py:166: error: Cannot determine type of 'CUM_FUNC_DEFAULTS'
    pandas\compat\numpy\function.py:167: error: Cannot determine type of 'CUM_FUNC_DEFAULTS'
    pandas\compat\numpy\function.py:169: error: Cannot determine type of 'CUM_FUNC_DEFAULTS'
    pandas\compat\numpy\function.py:188: error: Need type annotation for 'ALLANY_DEFAULTS'
    pandas\compat\numpy\function.py:189: error: Cannot determine type of 'ALLANY_DEFAULTS'
    pandas\compat\numpy\function.py:190: error: Cannot determine type of 'ALLANY_DEFAULTS'
    pandas\compat\numpy\function.py:191: error: Cannot determine type of 'ALLANY_DEFAULTS'
    pandas\compat\numpy\function.py:192: error: Cannot determine type of 'ALLANY_DEFAULTS'
    pandas\compat\numpy\function.py:194: error: Cannot determine type of 'ALLANY_DEFAULTS'
    pandas\compat\numpy\function.py:218: error: Need type annotation for 'SORT_DEFAULTS'
    pandas\compat\numpy\function.py:219: error: Cannot determine type of 'SORT_DEFAULTS'
    pandas\compat\numpy\function.py:220: error: Cannot determine type of 'SORT_DEFAULTS'
    pandas\compat\numpy\function.py:221: error: Cannot determine type of 'SORT_DEFAULTS'
    pandas\compat\numpy\function.py:222: error: Cannot determine type of 'SORT_DEFAULTS'
    pandas\compat\numpy\function.py:225: error: Need type annotation for 'STAT_FUNC_DEFAULTS'
    pandas\compat\numpy\function.py:226: error: Cannot determine type of 'STAT_FUNC_DEFAULTS'
    pandas\compat\numpy\function.py:227: error: Cannot determine type of 'STAT_FUNC_DEFAULTS'
    pandas\compat\numpy\function.py:229: error: Cannot determine type of 'STAT_FUNC_DEFAULTS'
    pandas\compat\numpy\function.py:233: error: Cannot determine type of 'STAT_FUNC_DEFAULTS'
    pandas\compat\numpy\function.py:237: error: Cannot determine type of 'STAT_FUNC_DEFAULTS'
    pandas\compat\numpy\function.py:239: error: Cannot determine type of 'STAT_FUNC_DEFAULTS'
    pandas\compat\numpy\function.py:245: error: Cannot determine type of 'STAT_FUNC_DEFAULTS'
    pandas\compat\numpy\function.py:250: error: Need type annotation for 'STAT_DDOF_FUNC_DEFAULTS'
    pandas\compat\numpy\function.py:251: error: Cannot determine type of 'STAT_DDOF_FUNC_DEFAULTS'
    pandas\compat\numpy\function.py:252: error: Cannot determine type of 'STAT_DDOF_FUNC_DEFAULTS'
    pandas\compat\numpy\function.py:253: error: Cannot determine type of 'STAT_DDOF_FUNC_DEFAULTS'
    pandas\compat\numpy\function.py:254: error: Cannot determine type of 'STAT_DDOF_FUNC_DEFAULTS'
    pandas\compat\numpy\function.py:257: error: Need type annotation for 'TAKE_DEFAULTS'
    pandas\compat\numpy\function.py:258: error: Cannot determine type of 'TAKE_DEFAULTS'
    pandas\compat\numpy\function.py:259: error: Cannot determine type of 'TAKE_DEFAULTS'
    pandas\compat\numpy\function.py:260: error: Cannot determine type of 'TAKE_DEFAULTS'

### PR DESCRIPTION
**Pull request by vaibhavhrt**
- [x] closes #25950 
- [ ] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

**Comment by vaibhavhrt** [comment]
@WillAyd I think this is my third open PR for typing. Should I wait for these to get reviewed before creating new issues and PR?

**Comment by WillAyd** [comment]
Fine by me - thanks for all of the help so far!

**Comment by vaibhavhrt** [comment]
Ok, I will wait for at least one of them to get merged before working on new issues.

File: pandas/compat/numpy/function.py
Comment (WillAyd) [inline pandas/compat/numpy/function.py]: Use Optional instead of Union None

File: pandas/compat/numpy/function.py
Comment (WillAyd) [inline pandas/compat/numpy/function.py]: Same

File: pandas/compat/numpy/function.py
Comment (WillAyd) [inline pandas/compat/numpy/function.py]: Seems strange to have a value of None-type - is this not populated elsewhere?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/compat/numpy/function.py
Comment (vaibhavhrt) [reply pandas/compat/numpy/function.py]: You mean like this, right?

    # type: OrderedDict[str, Optional[int, str]]

File: pandas/compat/numpy/function.py
Comment (vaibhavhrt) [reply pandas/compat/numpy/function.py]: No value of `CLIP_DEFAULTS` is never changed. There are two other similar variables at around Line 210-220.

File: pandas/compat/numpy/function.py
Comment (vaibhavhrt) [reply pandas/compat/numpy/function.py]: @WillAyd I changed all `Union` to `Optional`, where `None` was in there.

File: pandas/compat/numpy/function.py
Comment (WillAyd) [reply pandas/compat/numpy/function.py]: @gfyoung any insight here? Are these variables actually unused or is this intentional?

Comment (gfyoung) [reply pandas/compat/numpy/function.py]: `CLIP_DEFAULTS` --> `validate_clip` --> `validate_clip_with_axis` --> `clip` (in `core/generic.py`)

**Comment by gfyoung** [comment]
IMO, these type hints in this PR are a little over-fitted to this file.  These default argument dictionaries are meant to map strings to anything that `numpy` throws at us from its signatures upstream.  It would be more prudent to actual declare these types as `OrderedDict[str, Any]` to indicate this.

**Comment by WillAyd** [comment]
Yea was thinking the same thing, though I think maybe wrapping the containers in Final would be a good indication that these aren't meant to be modified:

https://mypy.readthedocs.io/en/latest/final_attrs.html?highlight=immutable#details-of-using-final

**Comment by gfyoung** [comment]
> though I think maybe wrapping the containers in Final would be a good indication that these aren't meant to be modified

That's a fair proposition.

**Comment by gfyoung** [comment]
My bad - fat fingers 😂

**Comment by vaibhavhrt** [comment]
@WillAyd @gfyoung does this looks good:

    CLIP_DEFAULTS = dict(out=None)  # type Final[Dict[str, Any]]

**Comment by gfyoung** [comment]
@vaibhavhrt : That looks good!

If you could make similar modifications to your other changes, that would be great.

**Comment by WillAyd** [comment]
@gfyoung think should convert dict to OrderedDict too? Not sure if a distinction between those is intentional so maybe make consistent?

**Comment by vaibhavhrt** [comment]
@WillAyd `Final` needs to be imported from `typing_extensions`(https://pypi.org/project/typing-extensions/). But I don't think we have that in requirements. I usually use pipenv so I am not sure how to add a new package to requirements in this project. Would it be enough to add `typing_extensions` to `requirements-dev.txt`

**Comment by WillAyd** [comment]
Well that's disappointing...not sure we want to add a third party library just yet. Was that not added until 3.7

**Comment by vaibhavhrt** [comment]
>mypy pandas/compat/numpy/function.py

runs without any error but at `from typing_extensions import Final` my linter says `unresolved import 'typing_extensions' Python(unresolved-import)`. I am guessing tests will fail.

**Comment by gfyoung** [comment]
> think should convert dict to OrderedDict too? Not sure if a distinction between those is intentional so maybe make consistent?

**Comment by WillAyd** [comment]
@jreback do you have any thoughts on adding `typing_extensions` as a dependency to pandas? 

https://github.com/python/typing/tree/master/typing_extensions

On the plus side maintained by Guido and could help smooth over compat issues leading up to 3.7 (where typing is no longer provisional). On the downside would be another dependency

From there I think we could use Final, ChainMap, OrderedDict, Type, etc...

Side note - I know we've used Type in a few places already but reading more up on it that wasn't added until 3.5.2 so I think that might have broken the code base for 3.5 and 3.5.1

File: pandas/compat/numpy/function.py
Comment (WillAyd) [inline pandas/compat/numpy/function.py]: Just a random callout especially for @gfyoung since we had a similar conversation yesterday, but I think there's a nuance to how we use the actual type of OrderedDict versus the generic type in the typing module (added recently in 3.7.2).

In this annotation it doesn't cause problems, but if this was in the signature of the function I think we would be getting a syntax error:

https://stackoverflow.com/questions/41207128/how-do-i-specify-ordereddict-k-v-types-for-mypy-type-annotation

Not sure we need to do anything yet just providing as FYI

Comment (gfyoung) [reply pandas/compat/numpy/function.py]: Indeed, that's true.  We'll have to keep an eye out for that as we slowly add more of the annotations to the codebase.

Comment (vaibhavhrt) [reply pandas/compat/numpy/function.py]: @WillAyd unrelated to typing but why do we take 3 lines to create and assign values to `CUM_FUNC_DEFAULTS`(and other similar variables) when it can be easily done in a single line. Upper case name indicates it's a constant, so shouldn't we do something like this:

    CUM_FUNC_DEFAULTS = OrderedDict(dtype=None, out=None)

Comment (gfyoung) [reply pandas/compat/numpy/function.py]: @vaibhavhrt : The 3 lines are to ensure that keys are inserted exactly in the order that we want them to.  If you use the one-liner, earlier Python 3.x versions would insert them based on Python's natural ordering of objects.

We probably could eventually replace with a one-liner, since later Python 3.x versions fix this issue, but I wouldn't do it immediately.

Comment (vaibhavhrt) [reply pandas/compat/numpy/function.py]: @gfyoung thanks for explaining.

**Comment by jreback** [comment]
not a big deal to add typing_extensions

it’s pretty common now-a-days

let’s do that in a separate patch

**Comment by vaibhavhrt** [comment]
Let me know when typing_extension has been added. @WillAyd if you want I can create an issue and PR to add typing_extensions to dependencies. But I am not sure how you are handling dependencies so  so I might need some help.

**Comment by WillAyd** [comment]
Sure give it a go! You should be able to see the PR to add Mypy which you can use for reference

**Comment by jbrockmendel** [comment]
@vaibhavhrt can you merge master

**Comment by vaibhavhrt** [comment]
@jbrockmendel ok I will merge master. I have few more commits to push in this PR. But before I do that #25975 needs to be merged.  The dependency `typing_extensions` is needed for this PR.  But that PR is having some problems with CI configuration. Can you take a look and tell me how to fix that.

**Comment by jreback** [comment]
@vaibhavhrt we won't be merging that PR mentioned https://github.com/pandas-dev/pandas/pull/25961#issuecomment-481106021 as is, see the comments

**Comment by vaibhavhrt** [comment]
Yeah I see. If no one is already working on adding required types to
pandas.typing I can do that in a separate  issue and PR.

File: pandas/compat/numpy/function.py
Comment (WillAyd) [inline pandas/compat/numpy/function.py]: Use Optional instead of Union[....,. None]

Comment (WillAyd) [reply pandas/compat/numpy/function.py]: This comment is applicable in a few other places

**Review by WillAyd — CHANGES_REQUESTED** [review]
Let's move forward without Final - can you merge master and address comments?

File: pandas/compat/numpy/function.py
Comment (vaibhavhrt) [reply pandas/compat/numpy/function.py]: @WillAyd , `Optional` takes only one argument, so instead of

    Union[X, Y, None]

Can I use optional like this 👇 

    Optional[Union[X, Y]]

I am a bit busy on weekdays, I will fix it if I get some time, otherwise it will be on saturday(friday for you), sorry about that.

**Comment by WillAyd** [comment]
Yes that’s correct.

File: pandas/compat/numpy/function.py
Comment (vaibhavhrt) [reply pandas/compat/numpy/function.py]: @WillAyd all done, turns out I found some free time. 😄

**Comment by vaibhavhrt** [comment]
@WillAyd two tests are failing, but only thing I did was add comments, how can it make tests fail.

**Review by WillAyd — APPROVED** [review]
lgtm. If you merge master one more time should resolve CI failures

**Comment by vaibhavhrt** [comment]
@jreback I have fixed everything, you can review and merge it.

**Comment by jreback** [comment]
thanks!

**Pull request merged**
